Tribler/community/doubleentry/database.py

```diff
@@ -0,0 +1,266 @@
+DATABASE_PATH = path.join(DATABASE_DIRECTORY, u"doubleentry.db")
+""" ID of the first block of the chain. """
+GENESIS_ID = "GENESIS_ID"
+""" Version to keep track if the db schema needs to be updated."""
+LATEST_DB_VERSION = 1
+""" Schema for the DoubleEntry DB."""
+schema = u"""
+CREATE TABLE IF NOT EXISTS double_entry(
+ block_hash			        text PRIMARY KEY,
+ previous_hash_requester	text NOT NULL,
+ public_key_requester		text NOT NULL,
+ signature_requester		text NOT NULL,
+ previous_hash_responder	text NOT NULL,
+ public_key_responder		text NOT NULL,
+ signature_responder		text NOT NULL,
+ sequence_number_requester  integer NOT NULL,
+ sequence_number_responder  integer NOT NULL,
+ up                         integer NOT NULL,
+ down                       integer NOT NULL,
+ total_up                   integer NOT NULL,
+ total_down                 integer NOT NULL
+);
+
+CREATE TABLE option(key TEXT PRIMARY KEY, value BLOB);
+INSERT INTO option(key, value) VALUES('database_version', '""" + str(LATEST_DB_VERSION) + u"""');
+INSERT INTO option(key, value) VALUES('previous_id', '""" + encode_db(GENESIS_ID) + u"""');
+"""
+
+cleanup = u"DELETE FROM double_entry;" \
+          u"UPDATE `option` SET value = '" + encode_db(GENESIS_ID) + \
+          u"' WHERE key == 'previous_id';"
+
+
+class Persistence:
+    """
+    Persistence layer for the DoubleEntry Community.
+    """
+    def __init__(self, working_directory):
+        """
+        Sets up the persistence layer ready for use.
+        :param working_directory: Path to the working directory
+        that will contain the the db at workingdirectory/DATABASE_PATH
+        :return:
+        """
+        self._working_directory = working_directory
+        self.db = None
+
+    def add_block(self, block_id, block):
+        """
+        Persist a block under a block_id
+        :param block_id: The ID the block is saved under. This is the hash of the block.
+        :param block: The data that will be saved.
+        """
+        self._init_database()
+        data = tuple(map(encode_db,
+                         (block_id,
+                          block.previous_hash_requester, block.public_key_requester, block.signature_requester,
+                          block.previous_hash_responder, block.public_key_responder, block.signature_responder,
+                          block.sequence_number_requester, block.sequence_number_responder,
+                          block.up, block.down, block.total_up, block.total_down)))
+
+        self.db.execute(
+            u"INSERT INTO double_entry (block_hash, previous_hash_requester, public_key_requester, "
+            u"signature_requester, previous_hash_responder, public_key_responder, signature_responder,"
+            u" sequence_number_requester, sequence_number_responder, up, down, total_up, total_down) "
+            u"values(?,?,?,?,?,?,?,?,?,?,?,?,?)",
+            data)
+
+        self._set_previous_id(block_id)
+        self.db.commit()
```

Comment:
Although, you would like to commit. This is not how Dispersy works. It will commit periodically.
Commiting yourself might cause Dispersy and your database to differ in content